pkg/scan/result.go

```diff
@@ -281,12 +281,16 @@ func (r *Results) SetSourceAndFilesystem(source string, f fs.FS, logicalSource b
 			newrng = defsecTypes.NewRangeWithLogicalSource(rng.GetLocalFilename(), rng.GetStartLine(), rng.GetEndLine(),
 				source, f)
 		}
+		parent := m.Parent()
 		switch {
 		case m.IsExplicit():
 			m = defsecTypes.NewExplicitMetadata(newrng, m.Reference())
 		default:
 			m = defsecTypes.NewMetadata(newrng, m.Reference())
 		}
+		if parent != nil {
+			m.SetParentPtr(parent)
```

Comment:
I'm unsure if this is sufficient/don't fully understand the use-case for this code. It may be that the metadata of the parent should also be cloned/overwritten here.

Reply:
What happens if you don't set the parent? Maybe we can try to add a test for this to observe and document the expected behaviour.

Reply:
If you don't set the parent the output will be missing the correct context of how the finding was reached, ex (from `tfsec` formatter):
```
──────────────────────────────────────────────────
  github.com/org/repo.git?ref=master/pwn.tf:11-13
──────────────────────────────────────────────────
```
Instead of the correct output:
```
──────────────────────────────────────────────────
  github.com/org/repo.git?ref=master/pwn.tf:11-13
   via test.tf:14-17 (module.example)
──────────────────────────────────────────────────
```
So the adding of the parent is necessary (one of the core features of this PR), it's more that this method (`SetSourceAndFilesystem`) is replacing the source/filesystem of the metadata in the results which maybe should include the parent metadata as well

Reply:
Right, I understand - I think there's value in testing this in isolation as well. At the moment the only tests we seem to have exercise this indirectly via the actual scanners, for example a call to: https://github.com/aquasecurity/defsec/blob/3ba2da0e8a11a7eff2e2f73fdc44d139791ac15d/pkg/scanners/dockerfile/scanner_test.go#L60 - will result in `SetSourceAnFilesystem` getting called. 

The other way to test this would be to add this assertion to each of the scanners, unless you've already done that in this PR, which I might have missed. 


What do you think?